File: frontend/lib/api/predictive.ts
import { api } from './client';

export interface PredictionExplanation {
  summary_text: string;
  key_drivers: Array<{ feature: string; value: any; impact: string }>;
  safety_notes: string[];
}

export interface PredictionOutcome {
  actual_numeric_outcome: number;
  outcome_label: string;
  error_magnitude?: number;
  recorded_by: string;
  recorded_at: string;
}

export interface PredictionRecord {
  id: string;
  tenant_id: string;
  model_id: string;
  prediction_type: string;
  entity_id: string;
  probability: number;
  risk_band: string;
  confidence_interval: { lower: number; upper: number };
  model_version: string;
  inference_timestamp: string;
  created_at: string;
  explanation?: PredictionExplanation;
  outcome?: PredictionOutcome;
}

export interface DecisionOverride {
  id: string;
  original_recommendation: string;
  chosen_action: string;
  override_reason: string;
  operator: string;
  created_at: string;
}

export interface DecisionSupportRecord {
  id: string;
  tenant_id: string;
  prediction_id: string;
  title: string;
  recommended_action: string;
  tradeoff_analysis: string;
  urgency: string;
  state: string;
  reviewed_by?: string;
  reviewed_at?: string;
  created_at: string;
  overrides?: DecisionOverride[];
}

export interface ModelGovernance {
  id: string;
  tenant_id: string;
  model_key: string;
  name: string;
  prediction_type: string;
  algorithm: string;
  current_version: string;
  status: string;
  thresholds: Record<string, any>;
  approved_by?: string;
  approved_at?: string;
  created_at: string;
}

export interface ForecastResponse {
  id: string;
  tenant_id: string;
  forecast_type: string;
  time_horizon: string;
  model_version: string;
  predictions_payload: Record<string, any>;
  confidence_intervals: Record<string, any>;
  created_at: string;
}

export interface ModelDriftEvent {
  id: string;
  tenant_id: string;
  model_key: string;
  drift_type: string;
  drift_metric_value: number;
  drift_status: string;
  details: Record<string, any>;
  detected_at: string;
}

export const predictiveApi = {
  listPredictions: (tenantId: string = 'default_tenant', predictionType?: string, riskBand?: string): Promise<PredictionRecord[]> => {
    const params = new URLSearchParams({ tenant_id: tenantId });
    if (predictionType) params.append('prediction_type', predictionType);
    if (riskBand) params.append('risk_band', riskBand);
    return api<PredictionRecord[]>(`/predictive/predictions?${params.toString()}`);
  },

  getPrediction: (predictionId: string): Promise<PredictionRecord> =>
    api<PredictionRecord>(`/predictive/predictions/${predictionId}`),

  generatePrediction: (data: { prediction_type: string; entity_id: string; features: Record<string, any> }, tenantId: string = 'default_tenant'): Promise<any> =>
    api<any>(`/predictive/predictions?tenant_id=${tenantId}`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  recordOutcome: (predictionId: string, data: { actual_numeric_outcome: number; outcome_label: string; recorded_by?: string }): Promise<any> =>
    api<any>(`/predictive/predictions/${predictionId}/outcome`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  listDecisionSupport: (tenantId: string = 'default_tenant', state?: string): Promise<DecisionSupportRecord[]> => {
    const params = new URLSearchParams({ tenant_id: tenantId });
    if (state) params.append('state', state);
    return api<DecisionSupportRecord[]>(`/predictive/decision-support?${params.toString()}`);
  },

  reviewDecisionSupport: (recordId: string, data: { action: string; reviewer?: string; notes?: string; override_reason?: string; chosen_action?: string }): Promise<DecisionSupportRecord> =>
    api<DecisionSupportRecord>(`/predictive/decision-support/${recordId}/review`, {
      method: 'POST',
      body: JSON.stringify(data),
    }),

  listModels: (tenantId: string = 'default_tenant'): Promise<ModelGovernance[]> =>
    api<ModelGovernance[]>(`/predictive/models?tenant_id=${tenantId}`),

  approveModel: (modelId: string, approvedBy: string = 'user'): Promise<ModelGovernance> =>
    api<ModelGovernance>(`/predictive/models/${modelId}/approve`, {
      method: 'POST',
      body: JSON.stringify({ approved_by: approvedBy }),
    }),

  getForecast: (tenantId: string = 'default_tenant', forecastType: string = 'WORKLOAD_DEMAND', timeHorizon: string = '30d'): Promise<any> =>
    api<any>(`/predictive/forecasts?tenant_id=${tenantId}&forecast_type=${forecastType}&time_horizon=${timeHorizon}`),

  getDriftSummary: (tenantId: string = 'default_tenant'): Promise<any> =>
    api<any>(`/predictive/model-monitoring/drift?tenant_id=${tenantId}`),
};

File: backend/app/schemas/predictive.py
"""Pydantic schemas for Phase 32: Advanced AI/ML Decision Intelligence & Predictive Operations API."""

from typing import Any, Dict, List, Optional
from datetime import datetime
from pydantic import BaseModel, Field


# =============================================================================
# 1. Predictions & Explanations Schemas
# =============================================================================

class PredictionExplanationSchema(BaseModel):
    summary_text: str
    key_drivers: List[Dict[str, Any]]
    safety_notes: List[str]

    class Config:
        from_attributes = True


class PredictionOutcomeSchema(BaseModel):
    actual_numeric_outcome: float
    outcome_label: str
    error_magnitude: Optional[float]
    recorded_by: str
    recorded_at: datetime

    class Config:
        from_attributes = True


class PredictionResponse(BaseModel):
    id: str
    tenant_id: str
    model_id: str
    prediction_type: str
    entity_id: str
    probability: float
    risk_band: str
    confidence_interval: Dict[str, float]
    model_version: str
    inference_timestamp: datetime
    created_at: datetime
    explanation: Optional[PredictionExplanationSchema] = None
    outcome: Optional[PredictionOutcomeSchema] = None

    class Config:
        from_attributes = True


class PredictionCreateRequest(BaseModel):
    prediction_type: str = Field(..., description="LEAD_CONVERSION, PROJECT_DELAY, PROJECT_EFFORT_VARIANCE, CLIENT_RETENTION")
    entity_id: str
    features: Dict[str, Any] = Field(default_factory=dict)


class OutcomeRecordRequest(BaseModel):
    actual_numeric_outcome: float = Field(..., description="1.0 for Won/Delayed or 0.0 for Lost/On-Time or actual hours")
    outcome_label: str = Field(..., description="e.g. 'CLOSED_WON', 'DELIVERED_ON_TIME'")
    recorded_by: str = Field(default="user")


# =============================================================================
# 2. Decision Support & Human Review Schemas
# =============================================================================

class DecisionOverrideSchema(BaseModel):
    id: str
    original_recommendation: str
    chosen_action: str
    override_reason: str
    operator: str
    created_at: datetime

    class Config:
        from_attributes = True


class DecisionSupportResponse(BaseModel):
    id: str
    tenant_id: str
    prediction_id: str
    title: str
    recommended_action: str
    tradeoff_analysis: str
    urgency: str
    state: str
    reviewed_by: Optional[str]
    reviewed_at: Optional[datetime]
    created_at: datetime
    overrides: List[DecisionOverrideSchema] = []

    class Config:
        from_attributes = True


class DecisionReviewRequest(BaseModel):
    action: str = Field(..., description="ACCEPT, OVERRIDE, REJECT")
    reviewer: str = Field(default="user")
    notes: Optional[str] = None
    override_reason: Optional[str] = None
    chosen_action: Optional[str] = None


# =============================================================================
# 3. Model Registry & Drift Monitoring Schemas
# =============================================================================

class ModelGovernanceResponse(BaseModel):
    id: str
    tenant_id: str
    model_key: str
    name: str
    prediction_type: str
    algorithm: str
    current_version: str
    status: str
    thresholds: Dict[str, Any]
    approved_by: Optional[str]
    approved_at: Optional[datetime]
    created_at: datetime

    class Config:
        from_attributes = True


class ModelApproveRequest(BaseModel):
    approved_by: str


class ForecastResponse(BaseModel):
    id: str
    tenant_id: str
    forecast_type: str
    time_horizon: str
    model_version: str
    predictions_payload: Dict[str, Any]
    confidence_intervals: Dict[str, Any]
    created_at: datetime

    class Config:
        from_attributes = True


class ModelDriftEventResponse(BaseModel):
    id: str
    tenant_id: str
    model_key: str
    drift_type: str
    drift_metric_value: float
    drift_status: str
    details: Dict[str, Any]
    detected_at: datetime

    class Config:
        from_attributes = True
